### PR TITLE
feat: add pnpm parser

### LINKED ISSUE #65: Support pnpm-lock.yaml
## Description
Parse pnpm-lock.yaml so as to extract package names/versions.
https://github.com/pnpm/spec/blob/master/lockfile/5.md

Example:
https://github.com/pnpm/pnpm/blob/main/pnpm-lock.yaml

## Issue
https://github.com/aquasecurity/trivy/issues/1200

### PR DESCRIPTION
**Pull request by mycodeself**
Hey, this PR is ~~a draft~~ for adding a pnpm parser. I know that there is other open but there seems to be no progress so I wanted to offer my solution. I hope it doesn't matter.

The issue is #65 

What I think is missing:

- ~~Add detection of direct / indirect dependencies~~
- ~~Add more tests~~

Doubts I have:

- Should the dev vulnerabilities be ignored?

If you think there is anything else I should consider, let me know. I want to try to finish it this week.

### CODE REVIEW

**Comment by mycodeself** [comment]
Hi, I have added more tests and added the detection of direct / indirect dependencies. 

Let me know what you think!

**Comment by knqyf263** [comment]
@mycodeself Thanks for your contribtuion!
@DmitriyLewen Could you take a look?

**Comment by DmitriyLewen** [comment]
@knqyf263  Looks good!
About `dev Dependencies`: I agree with @mycodeself - i think we don't need to show dev deps. 
Also i thought about `optional Dependencies`: if i understand correctly - these packages may not be installed and we will get a false positive.

**Comment by mycodeself** [comment]
@DmitriyLewen from [npm docs](https://docs.npmjs.com/cli/v8/configuring-npm/package-json#optionaldependencies) I understand that the `optionalDependencies` will be installed by default or at least attempt to install.

In fact, I have done some tests and as stated in the docs only when installing the dependencies with the `--no-optional` flag or when there is an error in fetching the dependency, the dependency is not installed.

Looking at the current parsers for nodejs:
- The `npm` parser will add this dependencies as currently is not filtering the optional dependencies and those are added in the `dependencies` object in the `package-lock.json`
- The `packagejson` parser will ignore the `optionalDependencies` as they are not added in the `dependencies` object in the `package.json` file
- The `yarn` parser will add this dependencies as they are added to the `yarn.lock` file

Therefore, if I am not mistaken, the optional dependencies are ignored only by the `packagejson`.

File: pkg/nodejs/pnpm/parse.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+	var lockFile LockFile
+	decoder := yaml.NewDecoder(r)
+	err := decoder.Decode(&lockFile)
+	if err != nil {
+		return nil, nil, xerrors.Errorf("decode error: %w", err)
+	}
+
+	libs, deps := p.parse(&lockFile)
+
+	return uniqueLibs(libs), uniqueDeps(deps), nil
```
Comment (knqyf263) [inline pkg/nodejs/pnpm/parse.go:61]: Do we need deduplication here? IIUC, we iterate `packages` that are unique.

Comment (mycodeself) [reply pkg/nodejs/pnpm/parse.go]: I have done some checking and you are correct! I have removed the unnecessary deduplication

let me know if is ok now :)

**Review by knqyf263 — APPROVED** [review]

**Comment by knqyf263** [comment]
Great work! Thanks for your contribution!

**Comment by knqyf263** [comment]
@mycodeself We recently deprecated fanal. Could you open a PR in Trivy?

**Comment by mycodeself** [comment]
@knqyf263 I'll try to take a look when I have some free time !

**Pull request merged**